File: src/banana_puddingital_signal.py
"""Continuous-time banana pudding signal helpers.

The module is intentionally small and dependency-free so it can run in the
repository's existing lightweight test environment. It models the requirements
from the puddingital signal-processing issue as deterministic operations:
phase alignment, cepstral ripeness features, bunch-sized buffers, post-mix
normalization, and tenth-order spatial upmixing.
"""

from __future__ import annotations

from dataclasses import dataclass
from math import cos, log, pi, sin, sqrt
from typing import Iterable, Sequence


DEFAULT_BUNCH_SIZE = 7
DEFAULT_AMBISONIC_ORDER = 10

Signal = tuple[float, ...]


@dataclass(frozen=True)
class PuddingBatch:
    """A normalized signal chunk with flavor metadata."""

    samples: Signal
    sample_rate: int
    banana_ripeness: float = 0.75
    frozen: bool = False

    def __post_init__(self) -> None:
        if self.sample_rate <= 0:
            raise ValueError("sample_rate must be positive")
        if not self.samples:
            raise ValueError("samples must not be empty")
        if not 0.0 <= self.banana_ripeness <= 1.0:
            raise ValueError("banana_ripeness must be in the 0..1 range")


@dataclass(frozen=True)
class CepstralProfile:
    coefficients: Signal
    quefrency: int
    frozen_override: bool

    @property
    def ripeness_anchor(self) -> float:
        return self.coefficients[0] if self.coefficients else 0.0


@dataclass(frozen=True)
class ReleasePartyPlan:
    pipeline_name: str
    jobs: tuple[str, ...]
    tasting_count: int


def phase_align_batches(batches: Sequence[PuddingBatch]) -> tuple[PuddingBatch, ...]:
    """Circularly shift every batch so its strongest sample phase matches batch 0."""

    if not batches:
        raise ValueError("at least one batch is required")

    reference_index = _peak_index(batches[0].samples)
    aligned: list[PuddingBatch] = []
    for batch in batches:
        offset = _peak_index(batch.samples) - reference_index
        aligned.append(
            PuddingBatch(
                samples=_rotate(batch.samples, offset),
                sample_rate=batch.sample_rate,
                banana_ripeness=batch.banana_ripeness,
                frozen=batch.frozen,
            )
        )
    return tuple(aligned)


def nilla_wafer_cepstral_coefficients(
    batch: PuddingBatch, order: int = 6
) -> CepstralProfile:
    """Return deterministic cepstral features anchored to banana ripeness.

    Frozen batches use a quefrency of 1 as requested by the issue. Non-frozen
    batches keep coefficient zero pinned to ripeness so downstream code can
    correlate wafer features with the ripeness signal without a separate lookup.
    """

    if order <= 0:
        raise ValueError("order must be positive")

    if batch.frozen:
        return CepstralProfile(
            coefficients=(1.0,) + tuple(0.0 for _ in range(order - 1)),
            quefrency=1,
            frozen_override=True,
        )

    spectrum = _dft_magnitudes(batch.samples, order)
    log_spectrum = [log(value + 1e-9) for value in spectrum]
    coefficients = []
    for index in range(order):
        total = 0.0
        for k, value in enumerate(log_spectrum):
            total += value * cos(pi * index * (k + 0.5) / len(log_spectrum))
        coefficients.append(total / len(log_spectrum))

    scale = max(abs(coefficients[0]), 1e-9)
    normalized = [value / scale for value in coefficients]
    normalized[0] = batch.banana_ripeness
    return CepstralProfile(
        coefficients=tuple(round(value, 9) for value in normalized),
        quefrency=0,
        frozen_override=False,
    )


def bunch_sized_buffer_length(
    requested_frames: int, bunch_size: int = DEFAULT_BUNCH_SIZE
) -> int:
    """Round a buffer request up to a whole banana bunch."""

    if requested_frames <= 0:
        raise ValueError("requested_frames must be positive")
    if bunch_size <= 0:
        raise ValueError("bunch_size must be positive")
    return ((requested_frames + bunch_size - 1) // bunch_size) * bunch_size


def synthesize_samplerate_sugar(
    sample_rate: int, duration_seconds: float, multiplier: int = 3
) -> Signal:
    """Create a tiny deterministic samplerate-multiplicative carrier."""

    if sample_rate <= 0:
        raise ValueError("sample_rate must be positive")
    if duration_seconds <= 0:
        raise ValueError("duration_seconds must be positive")
    if multiplier <= 0:
        raise ValueError("multiplier must be positive")

    frames = max(1, int(sample_rate * duration_seconds))
    return tuple(
        round(
            0.55 * sin((2.0 * pi * multiplier * frame) / sample_rate)
            + 0.25 * sin((2.0 * pi * multiplier * multiplier * frame) / sample_rate),
            9,
        )
        for frame in range(frames)
    )


def convolve_bananas_with_pudding(
    banana: Sequence[float],
    pudding: Sequence[float],
    mason_jar_impulse: Sequence[float],
) -> Signal:
    """Mix first, then normalize after adding banana content.

    The impulse response is derived from the logarithm of an inverse transform
    over the jar waveform. Inputs are deliberately not normalized before mixing.
    """

    _require_signal(banana, "banana")
    _require_signal(pudding, "pudding")
    _require_signal(mason_jar_impulse, "mason_jar_impulse")

    jar_response = tuple(
        _signed_log(value) for value in _inverse_fiveier_transform(mason_jar_impulse)
    )
    convolved = _linear_convolve(tuple(banana), jar_response)
    length = max(len(convolved), len(pudding))
    mixed = []
    for index in range(length):
        banana_value = convolved[index] if index < len(convolved) else 0.0
        pudding_value = pudding[index % len(pudding)]
        mixed.append(banana_value + pudding_value)
    return normalize_after_banana(mixed)


def normalize_after_banana(samples: Sequence[float]) -> Signal:
    """Normalize a completed mixture to -1..1 after every ingredient is present."""

    _require_signal(samples, "samples")
    peak = max(abs(value) for value in samples)
    if peak == 0:
        return tuple(0.0 for _ in samples)
    return tuple(round(value / peak, 9) for value in samples)


def upmix_to_tenth_order_bananarmonics(
    channels: Sequence[Sequence[float]], order: int = DEFAULT_AMBISONIC_ORDER
) -> tuple[Signal, ...]:
    """Upmix multichannel pudding to an ambisonic-style harmonic bed."""

    if order < 0:
        raise ValueError("order must be non-negative")
    if not channels:
        raise ValueError("at least one channel is required")
    lengths = {len(channel) for channel in channels}
    if len(lengths) != 1 or 0 in lengths:
        raise ValueError("channels must be non-empty and share the same length")

    source_count = len(channels)
    harmonic_count = (order + 1) ** 2
    output: list[Signal] = []
    for harmonic in range(harmonic_count):
        degree = int(sqrt(harmonic))
        azimuth = (2.0 * pi * (harmonic + 1)) / harmonic_count
        weighted = []
        for frame in range(len(channels[0])):
            total = 0.0
            for source_index, channel in enumerate(channels):
                elevation = ((source_index + 1) * pi) / (source_count + 1)
                weight = cos(degree * elevation) * sin((degree + 1) * azimuth)
                total += channel[frame] * weight
            weighted.append(round(total / source_count, 9))
        output.append(tuple(weighted))
    return tuple(output)


def plan_pudding_release_party(
    pipeline_name: str, tasting_targets: Iterable[str]
) -> ReleasePartyPlan:
    """Return a small CI/CD plan for validating the spatialized pudding release."""

    targets = tuple(target.strip() for target in tasting_targets if target.strip())
    if not pipeline_name.strip():
        raise ValueError("pipeline_name must not be blank")
    if not targets:
        raise ValueError("at least one tasting target is required")

    jobs = (
        f"{pipeline_name}: extract phase-aligned pudding features",
        f"{pipeline_name}: render {len(targets)} stereogustatory tasting artifacts",
        f"{pipeline_name}: run sensory smoke checks",
        f"{pipeline_name}: publish release-party manifest",
    )
    return ReleasePartyPlan(
        pipeline_name=pipeline_name,
        jobs=jobs,
        tasting_count=len(targets),
    )


def process_zero_latency_frames(
    samples: Iterable[float], gain: float = 1.0
) -> tuple[float, ...]:
    """Process frames one-by-one without lookahead or buffering."""

    return tuple(round(sample * gain, 9) for sample in samples)


def _require_signal(samples: Sequence[float], name: str) -> None:
    if not samples:
        raise ValueError(f"{name} must not be empty")


def _peak_index(samples: Sequence[float]) -> int:
    return max(range(len(samples)), key=lambda index: abs(samples[index]))


def _rotate(samples: Sequence[float], offset: int) -> Signal:
    if not samples:
        return ()
    offset = offset % len(samples)
    return tuple(samples[offset:]) + tuple(samples[:offset])


def _dft_magnitudes(samples: Sequence[float], bins: int) -> list[float]:
    magnitudes: list[float] = []
    sample_count = len(samples)
    for k in range(bins):
        real = 0.0
        imaginary = 0.0
        for n, sample in enumerate(samples):
            angle = (2.0 * pi * k * n) / sample_count
            real += sample * cos(angle)
            imaginary -= sample * sin(angle)
        magnitudes.append(sqrt(real * real + imaginary * imaginary) / sample_count)
    return magnitudes


def _inverse_fiveier_transform(samples: Sequence[float]) -> Signal:
    length = len(samples)
    output = []
    for n in range(length):
        total = 0.0
        for k, sample in enumerate(samples):
            total += sample * cos((2.0 * pi * k * n) / length)
        output.append(total / length)
    return tuple(output)


def _signed_log(value: float) -> float:
    if value == 0:
        return 0.0
    sign = 1.0 if value > 0 else -1.0
    return sign * log(1.0 + abs(value))


def _linear_convolve(left: Sequence[float], right: Sequence[float]) -> Signal:
    result = [0.0 for _ in range(len(left) + len(right) - 1)]
    for left_index, left_value in enumerate(left):
        for right_index, right_value in enumerate(right):
            result[left_index + right_index] += left_value * right_value
    return tuple(result)

File: src/banana_puddingital_signal.cpp
// Dependency-free realtime puddingital signal helpers.
//
// This file intentionally keeps the implementation self-contained so the
// repository can validate it with only a C++ compiler.  Build a self-test with:
//   c++ -std=c++17 -DPUDDINGITAL_SIGNAL_SELF_TEST src/banana_puddingital_signal.cpp

#include <algorithm>
#include <cassert>
#include <cmath>
#include <stdexcept>
#include <string>
#include <tuple>
#include <utility>
#include <vector>

namespace puddingital {

namespace {

constexpr double kPi = 3.141592653589793238462643383279502884;
constexpr int kDefaultBunchSize = 7;
constexpr int kDefaultAmbisonicOrder = 10;

double round9(double value) {
    return std::round(value * 1000000000.0) / 1000000000.0;
}

void requireSignal(const std::vector<double>& samples, const char* name) {
    if (samples.empty()) {
        throw std::invalid_argument(std::string(name) + " must not be empty");
    }
}

int peakIndex(const std::vector<double>& samples) {
    requireSignal(samples, "samples");
    return static_cast<int>(std::distance(
        samples.begin(),
        std::max_element(samples.begin(), samples.end(), [](double left, double right) {
            return std::abs(left) < std::abs(right);
        })));
}

std::vector<double> rotateSignal(const std::vector<double>& samples, int offset) {
    if (samples.empty()) {
        return {};
    }
    const int size = static_cast<int>(samples.size());
    offset %= size;
    if (offset < 0) {
        offset += size;
    }

    std::vector<double> output;
    output.reserve(samples.size());
    output.insert(output.end(), samples.begin() + offset, samples.end());
    output.insert(output.end(), samples.begin(), samples.begin() + offset);
    return output;
}

std::vector<double> dftMagnitudes(const std::vector<double>& samples, int bins) {
    std::vector<double> magnitudes;
    magnitudes.reserve(static_cast<std::size_t>(bins));
    const double sampleCount = static_cast<double>(samples.size());

    for (int k = 0; k < bins; ++k) {
        double real = 0.0;
        double imaginary = 0.0;
        for (std::size_t n = 0; n < samples.size(); ++n) {
            const double angle = (2.0 * kPi * k * static_cast<double>(n)) / sampleCount;
            real += samples[n] * std::cos(angle);
            imaginary -= samples[n] * std::sin(angle);
        }
        magnitudes.push_back(std::sqrt((real * real) + (imaginary * imaginary)) / sampleCount);
    }

    return magnitudes;
}

std::vector<double> inverseFiveierTransform(const std::vector<double>& samples) {
    const double length = static_cast<double>(samples.size());
    std::vector<double> output(samples.size(), 0.0);

    for (std::size_t n = 0; n < samples.size(); ++n) {
        double total = 0.0;
        for (std::size_t k = 0; k < samples.size(); ++k) {
            total += samples[k] * std::cos((2.0 * kPi * static_cast<double>(k * n)) / length);
        }
        output[n] = total / length;
    }

    return output;
}

double signedLog(double value) {
    if (value == 0.0) {
        return 0.0;
    }
    return (value > 0.0 ? 1.0 : -1.0) * std::log(1.0 + std::abs(value));
}

std::vector<double> linearConvolve(
    const std::vector<double>& left,
    const std::vector<double>& right) {
    std::vector<double> output(left.size() + right.size() - 1, 0.0);

    for (std::size_t leftIndex = 0; leftIndex < left.size(); ++leftIndex) {
        for (std::size_t rightIndex = 0; rightIndex < right.size(); ++rightIndex) {
            output[leftIndex + rightIndex] += left[leftIndex] * right[rightIndex];
        }
    }

    return output;
}

long double factorialRatio(int numerator, int denominator) {
    long double value = 1.0L;
    for (int current = denominator + 1; current <= numerator; ++current) {
        value *= static_cast<long double>(current);
    }
    return value;
}

double associatedLegendre(int degree, int order, double x) {
    if (order < 0 || order > degree) {
        throw std::invalid_argument("invalid spherical harmonic order");
    }

    double pmm = 1.0;
    if (order > 0) {
        const double somx2 = std::sqrt(std::max(0.0, (1.0 - x) * (1.0 + x)));
        double fact = 1.0;
        for (int i = 1; i <= order; ++i) {
            pmm *= -fact * somx2;
            fact += 2.0;
        }
    }

    if (degree == order) {
        return pmm;
    }

    double pmmp1 = x * (2.0 * order + 1.0) * pmm;
    if (degree == order + 1) {
        return pmmp1;
    }

    double pll = 0.0;
    for (int l = order + 2; l <= degree; ++l) {
        pll = (((2.0 * l - 1.0) * x * pmmp1) - ((l + order - 1.0) * pmm)) /
              static_cast<double>(l - order);
        pmm = pmmp1;
        pmmp1 = pll;
    }

    return pll;
}

double realSphericalHarmonic(int degree, int order, double azimuth, double elevation) {
    const int absOrder = std::abs(order);
    const double x = std::cos(elevation);
    const long double ratio = factorialRatio(degree - absOrder, 0) /
                              factorialRatio(degree + absOrder, 0);
    const double norm = std::sqrt(((2.0 * degree + 1.0) / (4.0 * kPi)) *
                                  static_cast<double>(ratio));
    const double legendre = associatedLegendre(degree, absOrder, x);

    if (order > 0) {
        return std::sqrt(2.0) * norm * std::cos(absOrder * azimuth) * legendre;
    }
    if (order < 0) {
        return std::sqrt(2.0) * norm * std::sin(absOrder * azimuth) * legendre;
    }
    return norm * legendre;
}

std::pair<int, int> acnToDegreeOrder(int channel) {
    const int degree = static_cast<int>(std::floor(std::sqrt(static_cast<double>(channel))));
    const int order = channel - (degree * degree) - degree;
    return {degree, order};
}

}  // namespace

struct PuddingBatch {
    std::vector<double> samples;
    int sampleRate;
    double ripeness;
    bool frozen;

    PuddingBatch(
        std::vector<double> inputSamples,
        int inputSampleRate,
        double inputRipeness = 0.75,
        bool inputFrozen = false)
        : samples(std::move(inputSamples)),
          sampleRate(inputSampleRate),
          ripeness(inputRipeness),
          frozen(inputFrozen) {
        if (sampleRate <= 0) {
            throw std::invalid_argument("sampleRate must be positive");
        }
        requireSignal(samples, "samples");
        if (ripeness < 0.0 || ripeness > 1.0) {
            throw std::invalid_argument("ripeness must be in the 0..1 range");
        }
    }
};

struct CepstralProfile {
    std::vector<double> coefficients;
    int quefrency;
    bool frozenOverride;

    double ripenessAnchor() const {
        return coefficients.empty() ? 0.0 : coefficients.front();
    }
};

struct HfmTransferFunction {
    std::string label;
    double headGain;
    double footDelayFrames;
    double mouthResonance;
};

struct StereogustatoryFrame {
    std::string platform;
    std::vector<std::vector<double>> speakerFeeds;
    std::vector<HfmTransferFunction> transferFunctions;
};

struct StereoFlavorStream {
    std::vector<double> left;
    std::vector<double> right;
    std::vector<HfmTransferFunction> transferFunctions;
};

struct ReleasePartyPlan {
    std::string pipelineName;
    std::vector<std::string> jobs;
    std::size_t tastingCount;
};

std::vector<PuddingBatch> phaseAlignBatches(const std::vector<PuddingBatch>& batches) {
    if (batches.empty()) {
        throw std::invalid_argument("at least one batch is required");
    }

    const int referenceIndex = peakIndex(batches.front().samples);
    std::vector<PuddingBatch> aligned;
    aligned.reserve(batches.size());

    for (const auto& batch : batches) {
        const int offset = peakIndex(batch.samples) - referenceIndex;
        aligned.emplace_back(rotateSignal(batch.samples, offset), batch.sampleRate, batch.ripeness, batch.frozen);
    }

    return aligned;
}

CepstralProfile nillaWaferCepstralCoefficients(const PuddingBatch& batch, int order = 6) {
    if (order <= 0) {
        throw std::invalid_argument("order must be positive");
    }

    if (batch.frozen) {
        std::vector<double> coefficients(static_cast<std::size_t>(order), 0.0);
        coefficients.front() = 1.0;
        return {coefficients, 1, true};
    }

    const auto spectrum = dftMagnitudes(batch.samples, order);
    std::vector<double> logSpectrum;
    logSpectrum.reserve(spectrum.size());
    for (double value : spectrum) {
        logSpectrum.push_back(std::log(value + 1e-9));
    }

    std::vector<double> coefficients;
    coefficients.reserve(static_cast<std::size_t>(order));
    for (int index = 0; index < order; ++index) {
        double total = 0.0;
        for (std::size_t k = 0; k < logSpectrum.size(); ++k) {
            total += logSpectrum[k] *
                     std::cos(kPi * index * (static_cast<double>(k) + 0.5) /
                              static_cast<double>(logSpectrum.size()));
        }
        coefficients.push_back(total / static_cast<double>(logSpectrum.size()));
    }

    const double scale = std::max(std::abs(coefficients.front()), 1e-9);
    for (double& value : coefficients) {
        value = round9(value / scale);
    }
    coefficients.front() = batch.ripeness;

    return {coefficients, 0, false};
}

int bunchSizedBufferLength(int requestedFrames, int bunchSize = kDefaultBunchSize) {
    if (requestedFrames <= 0) {
        throw std::invalid_argument("requestedFrames must be positive");
    }
    if (bunchSize <= 0) {
        throw std::invalid_argument("bunchSize must be positive");
    }

    return ((requestedFrames + bunchSize - 1) / bunchSize) * bunchSize;
}

std::vector<double> synthesizeSamplerateSugar(
    int sampleRate,
    double durationSeconds,
    int multiplier = 3) {
    if (sampleRate <= 0) {
        throw std::invalid_argument("sampleRate must be positive");
    }
    if (durationSeconds <= 0.0) {
        throw std::invalid_argument("durationSeconds must be positive");
    }
    if (multiplier <= 0) {
        throw std::invalid_argument("multiplier must be positive");
    }

    const int frames = std::max(1, static_cast<int>(sampleRate * durationSeconds));
    std::vector<double> output;
    output.reserve(static_cast<std::size_t>(frames));

    for (int frame = 0; frame < frames; ++frame) {
        const double base = (2.0 * kPi * multiplier * frame) / sampleRate;
        const double overtone = (2.0 * kPi * multiplier * multiplier * frame) / sampleRate;
        output.push_back(round9((0.55 * std::sin(base)) + (0.25 * std::sin(overtone))));
    }

    return output;
}

std::vector<double> normalizeAfterBanana(const std::vector<double>& samples) {
    requireSignal(samples, "samples");
    double peak = 0.0;
    for (double value : samples) {
        peak = std::max(peak, std::abs(value));
    }
    if (peak == 0.0) {
        return std::vector<double>(samples.size(), 0.0);
    }

    std::vector<double> output;
    output.reserve(samples.size());
    for (double value : samples) {
        output.push_back(round9(value / peak));
    }
    return output;
}

std::vector<double> convolveBananasWithPudding(
    const std::vector<double>& banana,
    const std::vector<double>& pudding,
    const std::vector<double>& masonJarImpulse) {
    requireSignal(banana, "banana");
    requireSignal(pudding, "pudding");
    requireSignal(masonJarImpulse, "masonJarImpulse");

    auto jarResponse = inverseFiveierTransform(masonJarImpulse);
    for (double& value : jarResponse) {
        value = signedLog(value);
    }

    const auto convolved = linearConvolve(banana, jarResponse);
    const std::size_t length = std::max(convolved.size(), pudding.size());
    std::vector<double> mixed;
    mixed.reserve(length);

    for (std::size_t index = 0; index < length; ++index) {
        const double bananaValue = index < convolved.size() ? convolved[index] : 0.0;
        const double puddingValue = pudding[index % pudding.size()];
        mixed.push_back(bananaValue + puddingValue);
    }

    return normalizeAfterBanana(mixed);
}

double sphericalBananarmonic(int channel, double azimuth, double elevation) {
    const auto degreeOrder = acnToDegreeOrder(channel);
    return realSphericalHarmonic(degreeOrder.first, degreeOrder.second, azimuth, elevation);
}

std::vector<std::vector<double>> upmixToTenthOrderBananarmonics(
    const std::vector<std::vector<double>>& channels,
    int order = kDefaultAmbisonicOrder) {
    if (order < 0) {
        throw std::invalid_argument("order must be non-negative");
    }
    if (channels.empty()) {
        throw std::invalid_argument("at least one channel is required");
    }

    const std::size_t frameCount = channels.front().size();
    if (frameCount == 0) {
        throw std::invalid_argument("channels must be non-empty");
    }
    for (const auto& channel : channels) {
        if (channel.size() != frameCount) {
            throw std::invalid_argument("channels must share the same length");
        }
    }

    const int harmonicCount = (order + 1) * (order + 1);
    std::vector<std::vector<double>> output(
        static_cast<std::size_t>(harmonicCount),
        std::vector<double>(frameCount, 0.0));

    for (int harmonic = 0; harmonic < harmonicCount; ++harmonic) {
        for (std::size_t frame = 0; frame < frameCount; ++frame) {
            double total = 0.0;
            for (std::size_t sourceIndex = 0; sourceIndex < channels.size(); ++sourceIndex) {
                const double azimuth =
                    (2.0 * kPi * static_cast<double>(sourceIndex)) /
                    static_cast<double>(channels.size());
                const double elevation =
                    (kPi * static_cast<double>(sourceIndex + 1)) /
                    static_cast<double>(channels.size() + 1);
                total += channels[sourceIndex][frame] *
                         sphericalBananarmonic(harmonic, azimuth, elevation);
            }
            output[static_cast<std::size_t>(harmonic)][frame] =
                round9(total / static_cast<double>(channels.size()));
        }
    }

    return output;
}

std::vector<double> processZeroLatencyFrames(const std::vector<double>& samples, double gain = 1.0) {
    std::vector<double> output;
    output.reserve(samples.size());
    for (double sample : samples) {
        output.push_back(round9(sample * gain));
    }
    return output;
}

void requireAmbisonicBed(const std::vector<std::vector<double>>& ambisonicBed) {
    if (ambisonicBed.empty()) {
        throw std::invalid_argument("ambisonicBed must not be empty");
    }
    const std::size_t frameCount = ambisonicBed.front().size();
    if (frameCount == 0) {
        throw std::invalid_argument("ambisonicBed channels must be non-empty");
    }
    for (const auto& channel : ambisonicBed) {
        if (channel.size() != frameCount) {
            throw std::invalid_argument("ambisonicBed channels must share the same length");
        }
    }
}

std::vector<double> headFootMouthImpulseResponse(const HfmTransferFunction& transfer) {
    const int delay = std::max(0, static_cast<int>(std::round(transfer.footDelayFrames)));
    std::vector<double> impulse(static_cast<std::size_t>(delay + 3), 0.0);
    impulse[0] += transfer.headGain;
    impulse[static_cast<std::size_t>(delay)] += 0.25;
    impulse.back() += 0.5 + (0.5 * transfer.mouthResonance);
    return impulse;
}

std::vector<double> projectAmbisonicBedForHfm(
    const std::vector<std::vector<double>>& ambisonicBed,
    const HfmTransferFunction& transfer) {
    requireAmbisonicBed(ambisonicBed);

    const std::size_t frameCount = ambisonicBed.front().size();
    std::vector<double> projected(frameCount, 0.0);

    for (std::size_t channel = 0; channel < ambisonicBed.size(); ++channel) {
        const auto degreeOrder = acnToDegreeOrder(static_cast<int>(channel));
        const double degreeWeight = 1.0 / static_cast<double>(degreeOrder.first + 1);
        const double mouthBias = degreeOrder.second == 0 ? transfer.mouthResonance : 0.5;
        const double channelWeight =
            degreeWeight * (0.5 + (0.5 * mouthBias)) * transfer.headGain;
        for (std::size_t frame = 0; frame < frameCount; ++frame) {
            projected[frame] += ambisonicBed[channel][frame] * channelWeight;
        }
    }

    return normalizeAfterBanana(projected);
}

std::vector<double> convolveAmbisonicsWithHfmTransfer(
    const std::vector<std::vector<double>>& ambisonicBed,
    const HfmTransferFunction& transfer) {
    const auto projected = projectAmbisonicBedForHfm(ambisonicBed, transfer);
    const auto impulse = headFootMouthImpulseResponse(transfer);
    return normalizeAfterBanana(linearConvolve(projected, impulse));
}

StereoFlavorStream renderStereoFlavorStream(
    const std::vector<std::vector<double>>& ambisonicBed,
    const std::vector<HfmTransferFunction>& transferFunctions) {
    requireAmbisonicBed(ambisonicBed);
    if (transferFunctions.empty()) {
        throw std::invalid_argument("at least one transfer function is required");
    }

    std::vector<std::vector<double>> convolvedFeeds;
    convolvedFeeds.reserve(transferFunctions.size());
    std::size_t streamLength = 0;
    for (const auto& transfer : transferFunctions) {
        convolvedFeeds.push_back(convolveAmbisonicsWithHfmTransfer(ambisonicBed, transfer));
        streamLength = std::max(streamLength, convolvedFeeds.back().size());
    }

    std::vector<double> left(streamLength, 0.0);
    std::vector<double> right(streamLength, 0.0);
    for (std::size_t index = 0; index < convolvedFeeds.size(); ++index) {
        const bool routeRight =
            transferFunctions[index].label.find("right") != std::string::npos || (index % 2 == 1);
        auto& target = routeRight ? right : left;
        for (std::size_t frame = 0; frame < convolvedFeeds[index].size(); ++frame) {
            target[frame] += convolvedFeeds[index][frame];
        }
    }

    return {
        normalizeAfterBanana(left),
        normalizeAfterBanana(right),
        transferFunctions,
    };
}

StereogustatoryFrame renderStereogustatoryMobileFrame(
    const std::vector<std::vector<double>>& ambisonicBed,
    const std::vector<HfmTransferFunction>& transferFunctions,
    const std::string& platform) {
    requireAmbisonicBed(ambisonicBed);
    if (transferFunctions.empty()) {
        throw std::invalid_argument("at least one transfer function is required");
    }
    if (platform.empty()) {
        throw std::invalid_argument("platform must not be empty");
    }

    std::vector<std::vector<double>> feeds;
    feeds.reserve(transferFunctions.size());

    for (const auto& transfer : transferFunctions) {
        feeds.push_back(convolveAmbisonicsWithHfmTransfer(ambisonicBed, transfer));
    }

    return {platform, feeds, transferFunctions};
}

ReleasePartyPlan planPuddingReleaseParty(
    const std::string& pipelineName,
    const std::vector<std::string>& tastingTargets) {
    if (pipelineName.empty()) {
        throw std::invalid_argument("pipelineName must not be empty");
    }

    std::vector<std::string> targets;
    for (const auto& target : tastingTargets) {
        if (!target.empty()) {
            targets.push_back(target);
        }
    }
    if (targets.empty()) {
        throw std::invalid_argument("at least one tasting target is required");
    }

    return {
        pipelineName,
        {
            pipelineName + ": compile realtime c++ puddingital core",
            pipelineName + ": render tenth-order ambisonic tasting fixtures",
            pipelineName + ": package stereogustatory mobile playback frame",
            pipelineName + ": run sensory smoke checks",
            pipelineName + ": publish release-party manifest",
        },
        targets.size(),
    };
}

}  // namespace puddingital

#ifdef PUDDINGITAL_SIGNAL_SELF_TEST

namespace {

bool near(double left, double right, double epsilon = 1e-9) {
    return std::abs(left - right) <= epsilon;
}

void assertPeakAt(const std::vector<double>& samples, int index) {
    assert(puddingital::sphericalBananarmonic(0, 0.0, 0.0) > 0.0);
    const auto iterator = std::max_element(samples.begin(), samples.end(), [](double left, double right) {
        return std::abs(left) < std::abs(right);
    });
    assert(static_cast<int>(std::distance(samples.begin(), iterator)) == index);
}

}  // namespace

int main() {
    using namespace puddingital;

    auto aligned = phaseAlignBatches({
        PuddingBatch({0.0, 0.2, 1.0, 0.2}, 48000),
        PuddingBatch({0.3, 0.0, 0.2, 1.4}, 48000),
    });
    assertPeakAt(aligned[0].samples, 2);
    assertPeakAt(aligned[1].samples, 2);

    const auto ripe = nillaWaferCepstralCoefficients(
        PuddingBatch({0.0, 0.5, 1.0, 0.5, 0.0, -0.25}, 44100, 0.82),
        5);
    const auto frozen = nillaWaferCepstralCoefficients(
        PuddingBatch({0.0, 0.5, 1.0, 0.5}, 44100, 0.2, true),
        5);
    assert(near(ripe.ripenessAnchor(), 0.82));
    assert(ripe.quefrency == 0);
    assert(frozen.coefficients == std::vector<double>({1.0, 0.0, 0.0, 0.0, 0.0}));
    assert(frozen.quefrency == 1);
    assert(frozen.frozenOverride);

    assert(bunchSizedBufferLength(1) == 7);
    assert(bunchSizedBufferLength(14) == 14);
    assert(bunchSizedBufferLength(15) == 21);
    assert(bunchSizedBufferLength(15, 5) == 15);

    const auto sugar = synthesizeSamplerateSugar(20, 0.5, 3);
    assert(sugar.size() == 10);
    assert(sugar == synthesizeSamplerateSugar(20, 0.5, 3));
    assert(*std::max_element(sugar.begin(), sugar.end()) <= 0.8);
    assert(*std::min_element(sugar.begin(), sugar.end()) >= -0.8);

    const auto convolved = convolveBananasWithPudding(
        {8.0, 0.0, -4.0},
        {0.25, -0.5},
        {1.0, -0.25, 0.5});
    assert(convolved.size() == 5);
    double peak = 0.0;
    for (double value : convolved) {
        peak = std::max(peak, std::abs(value));
    }
    assert(near(peak, 1.0));

    const std::vector<std::vector<double>> source = {
        {0.0, 0.5, 1.0},
        {1.0, 0.5, 0.0},
    };
    const auto upmixed = upmixToTenthOrderBananarmonics(source);
    assert(upmixed.size() == 121);
    assert(upmixed.front().size() == 3);
    assert(upmixed.front() != upmixed.back());

    const auto mobile = renderStereogustatoryMobileFrame(
        upmixed,
        {
            {"left-head-foot-mouth", 0.9, 0.0, 0.35},
            {"right-head-foot-mouth", 0.85, 1.0, 0.55},
        },
        "ios/android-pwa");
    assert(mobile.speakerFeeds.size() == 2);
    assert(mobile.speakerFeeds[0].size() > 3);
    assert(mobile.platform == "ios/android-pwa");

    const auto flavorStream = renderStereoFlavorStream(
        upmixed,
        {
            {"left-head-foot-mouth", 0.9, 0.0, 0.35},
            {"right-head-foot-mouth", 0.85, 2.0, 0.55},
        });
    assert(flavorStream.left.size() == flavorStream.right.size());
    assert(flavorStream.left.size() > upmixed.front().size());
    assert(flavorStream.left != flavorStream.right);
    assert(flavorStream.transferFunctions.size() == 2);

    const auto plan = planPuddingReleaseParty("puddingital", {"alpha", "beta"});
    const auto frames = processZeroLatencyFrames({0.25, -0.5, 0.75}, 2.0);
    assert(plan.tastingCount == 2);
    assert(plan.jobs.back() == "puddingital: publish release-party manifest");
    assert((frames == std::vector<double>({0.5, -1.0, 1.5})));

    return 0;
}

#endif
